File: articles/conservation-biology-protection-of-life/python/conservation_priority.py
"""
Conservation priority scoring.

This script reads a small conservation-unit dataset and calculates priority
scores using transparent weights. It also performs a simple sensitivity check.

Run:
    python python/conservation_priority.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "conservation_units.csv"


def score_units(data: pd.DataFrame, weights: dict[str, float]) -> pd.DataFrame:
    """Calculate weighted conservation priority scores."""

    scored = data.copy()
    scored["priority_score"] = sum(
        weights[column] * scored[column] for column in weights
    )

    return scored.sort_values("priority_score", ascending=False).reset_index(drop=True)


def main() -> None:
    """Run default and alternative scoring frameworks."""

    units = pd.read_csv(DATA_PATH)

    default_weights = {
        "extinction_risk": 0.30,
        "endemism": 0.20,
        "habitat_loss": 0.20,
        "fragmentation": 0.15,
        "recovery_potential": 0.10,
        "cost_index": -0.05,
    }

    alternative_weights = default_weights.copy()
    alternative_weights["recovery_potential"] = 0.20
    alternative_weights["fragmentation"] = 0.10

    default_ranked = score_units(units, default_weights)
    alternative_ranked = score_units(units, alternative_weights)

    comparison = default_ranked[["unit", "priority_score"]].merge(
        alternative_ranked[["unit", "priority_score"]],
        on="unit",
        suffixes=("_default", "_alternative"),
    )

    comparison["rank_default"] = comparison["priority_score_default"].rank(
        ascending=False,
        method="min",
    )
    comparison["rank_alternative"] = comparison["priority_score_alternative"].rank(
        ascending=False,
        method="min",
    )

    print(comparison.round(3).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/conservation-biology-protection-of-life/go/conservation_priority.go
// Conservation priority scoring in Go.
//
// This portable example demonstrates a simple weighted conservation scoring
// workflow without external dependencies.

package main

import (
	"fmt"
	"sort"
)

type ConservationUnit struct {
	Name              string
	ExtinctionRisk    float64
	Endemism          float64
	HabitatLoss       float64
	Fragmentation     float64
	RecoveryPotential float64
	CostIndex         float64
	PriorityScore     float64
}

func scoreUnit(unit ConservationUnit) float64 {
	return 0.30*unit.ExtinctionRisk +
		0.20*unit.Endemism +
		0.20*unit.HabitatLoss +
		0.15*unit.Fragmentation +
		0.10*unit.RecoveryPotential -
		0.05*unit.CostIndex
}

func main() {
	units := []ConservationUnit{
		{"A", 0.92, 0.80, 0.75, 0.88, 0.45, 0.60, 0.0},
		{"B", 0.65, 0.30, 0.90, 0.70, 0.80, 0.45, 0.0},
		{"C", 0.40, 0.25, 0.35, 0.40, 0.70, 0.30, 0.0},
		{"D", 0.85, 0.95, 0.60, 0.92, 0.35, 0.75, 0.0},
		{"E", 0.55, 0.50, 0.70, 0.60, 0.65, 0.50, 0.0},
	}

	for i := range units {
		units[i].PriorityScore = scoreUnit(units[i])
	}

	sort.Slice(units, func(i, j int) bool {
		return units[i].PriorityScore > units[j].PriorityScore
	})

	for _, unit := range units {
		fmt.Printf("unit=%s priority_score=%.3f\n", unit.Name, unit.PriorityScore)
	}
}
